<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.5.2" />
<title>Just Annotate My Genome (JAMg) info</title>
<style type="text/css">
/* Debug borders */
p, li, dt, dd, div, pre, h1, h2, h3, h4, h5, h6 {
/*
  border: 1px solid red;
*/
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

tt {
  color: navy;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  font-family: sans-serif;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}

div.sectionbody {
  font-family: serif;
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}

pre {
  padding: 0;
  margin: 0;
}

span#author {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  font-size: 1.1em;
}
span#email {
}
span#revnumber, span#revdate, span#revremark {
  font-family: sans-serif;
}

div#footer {
  font-family: sans-serif;
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
div#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
div#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

div#preamble {
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.tableblock, div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.0em;
  margin-bottom: 2.0em;
  margin-right: 10%;
  color: #606060;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}

div.listingblock > div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock, div.verseblock {
  padding-left: 1.0em;
  margin-left: 1.0em;
  margin-right: 10%;
  border-left: 5px solid #dddddd;
  color: #777777;
}

div.quoteblock > div.attribution {
  padding-top: 0.5em;
  text-align: right;
}

div.verseblock > div.content {
  white-space: pre;
}
div.verseblock > div.attribution {
  padding-top: 0.75em;
  text-align: left;
}
/* DEPRECATED: Pre version 8.2.7 verse style literal block. */
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 3px solid #dddddd;
}

div.exampleblock > div.content {
  border-left: 3px solid #dddddd;
  padding-left: 0.5em;
}

div.imageblock div.content { padding-left: 0; }
span.image img { border-style: none; }
a.image:visited { color: white; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
  color: navy;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
ol.arabic {
  list-style-type: decimal;
}
ol.loweralpha {
  list-style-type: lower-alpha;
}
ol.upperalpha {
  list-style-type: upper-alpha;
}
ol.lowerroman {
  list-style-type: lower-roman;
}
ol.upperroman {
  list-style-type: upper-roman;
}

div.compact ul, div.compact ol,
div.compact p, div.compact p,
div.compact div, div.compact div {
  margin-top: 0.1em;
  margin-bottom: 0.1em;
}

div.tableblock > table {
  border: 3px solid #527bbd;
}
thead, p.table.header {
  font-family: sans-serif;
  font-weight: bold;
}
tfoot {
  font-weight: bold;
}
td > div.verse {
  white-space: pre;
}
p.table {
  margin-top: 0;
}
/* Because the table frame attribute is overriden by CSS in most browsers. */
div.tableblock > table[frame="void"] {
  border-style: none;
}
div.tableblock > table[frame="hsides"] {
  border-left-style: none;
  border-right-style: none;
}
div.tableblock > table[frame="vsides"] {
  border-top-style: none;
  border-bottom-style: none;
}


div.hdlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hdlist tr {
  padding-bottom: 15px;
}
dt.hdlist1.strong, td.hdlist1.strong {
  font-weight: bold;
}
td.hdlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
  color: navy;
}
td.hdlist2 {
  vertical-align: top;
}
div.hdlist.compact tr {
  margin: 0;
  padding-bottom: 0;
}

.comment {
  background: yellow;
}

.footnote, .footnoteref {
  font-size: 0.8em;
}

span.footnote, span.footnoteref {
  vertical-align: super;
}

#footnotes {
  margin: 20px 0 20px 0;
  padding: 7px 0 0 0;
}

#footnotes div.footnote {
  margin: 0 0 5px 0;
}

#footnotes hr {
  border: none;
  border-top: 1px solid silver;
  height: 1px;
  text-align: left;
  margin-left: 0;
  width: 20%;
  min-width: 100px;
}


@media print {
  div#footer-badges { display: none; }
}

div#toc {
  margin-bottom: 2.5em;
}

div#toctitle {
  color: #527bbd;
  font-family: sans-serif;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}
/* Workarounds for IE6's broken and incomplete CSS2. */

div.sidebar-content {
  background: #ffffee;
  border: 1px solid silver;
  padding: 0.5em;
}
div.sidebar-title, div.image-title {
  color: #527bbd;
  font-family: sans-serif;
  font-weight: bold;
  margin-top: 0.0em;
  margin-bottom: 0.5em;
}

div.listingblock div.content {
  border: 1px solid silver;
  background: #f4f4f4;
  padding: 0.5em;
}

div.quoteblock-attribution {
  padding-top: 0.5em;
  text-align: right;
}

div.verseblock-content {
  white-space: pre;
}
div.verseblock-attribution {
  padding-top: 0.75em;
  text-align: left;
}

div.exampleblock-content {
  border-left: 3px solid #dddddd;
  padding-left: 0.5em;
}

/* IE6 sets dynamically generated links as visited. */
div#toc a:visited { color: blue; }
</style>
<script type="text/javascript">
/*<![CDATA[*/
window.onload = function(){asciidoc.footnotes();}
var asciidoc = {  // Namespace.

/////////////////////////////////////////////////////////////////////
// Table Of Contents generator
/////////////////////////////////////////////////////////////////////

/* Author: Mihai Bazon, September 2002
 * http://students.infoiasi.ro/~mishoo
 *
 * Table Of Content generator
 * Version: 0.4
 *
 * Feel free to use this script under the terms of the GNU General Public
 * License, as long as you do not remove or alter this notice.
 */

 /* modified by Troy D. Hanson, September 2006. License: GPL */
 /* modified by Stuart Rackham, 2006, 2009. License: GPL */

// toclevels = 1..4.
toc: function (toclevels) {

  function getText(el) {
    var text = "";
    for (var i = el.firstChild; i != null; i = i.nextSibling) {
      if (i.nodeType == 3 /* Node.TEXT_NODE */) // IE doesn't speak constants.
        text += i.data;
      else if (i.firstChild != null)
        text += getText(i);
    }
    return text;
  }

  function TocEntry(el, text, toclevel) {
    this.element = el;
    this.text = text;
    this.toclevel = toclevel;
  }

  function tocEntries(el, toclevels) {
    var result = new Array;
    var re = new RegExp('[hH]([2-'+(toclevels+1)+'])');
    // Function that scans the DOM tree for header elements (the DOM2
    // nodeIterator API would be a better technique but not supported by all
    // browsers).
    var iterate = function (el) {
      for (var i = el.firstChild; i != null; i = i.nextSibling) {
        if (i.nodeType == 1 /* Node.ELEMENT_NODE */) {
          var mo = re.exec(i.tagName);
          if (mo && (i.getAttribute("class") || i.getAttribute("className")) != "float") {
            result[result.length] = new TocEntry(i, getText(i), mo[1]-1);
          }
          iterate(i);
        }
      }
    }
    iterate(el);
    return result;
  }

  var toc = document.getElementById("toc");
  var entries = tocEntries(document.getElementById("content"), toclevels);
  for (var i = 0; i < entries.length; ++i) {
    var entry = entries[i];
    if (entry.element.id == "")
      entry.element.id = "_toc_" + i;
    var a = document.createElement("a");
    a.href = "#" + entry.element.id;
    a.appendChild(document.createTextNode(entry.text));
    var div = document.createElement("div");
    div.appendChild(a);
    div.className = "toclevel" + entry.toclevel;
    toc.appendChild(div);
  }
  if (entries.length == 0)
    toc.parentNode.removeChild(toc);
},


/////////////////////////////////////////////////////////////////////
// Footnotes generator
/////////////////////////////////////////////////////////////////////

/* Based on footnote generation code from:
 * http://www.brandspankingnew.net/archive/2005/07/format_footnote.html
 */

footnotes: function () {
  var cont = document.getElementById("content");
  var noteholder = document.getElementById("footnotes");
  var spans = cont.getElementsByTagName("span");
  var refs = {};
  var n = 0;
  for (i=0; i<spans.length; i++) {
    if (spans[i].className == "footnote") {
      n++;
      // Use [\s\S] in place of . so multi-line matches work.
      // Because JavaScript has no s (dotall) regex flag.
      note = spans[i].innerHTML.match(/\s*\[([\s\S]*)]\s*/)[1];
      noteholder.innerHTML +=
        "<div class='footnote' id='_footnote_" + n + "'>" +
        "<a href='#_footnoteref_" + n + "' title='Return to text'>" +
        n + "</a>. " + note + "</div>";
      spans[i].innerHTML =
        "[<a id='_footnoteref_" + n + "' href='#_footnote_" + n +
        "' title='View footnote' class='footnote'>" + n + "</a>]";
      var id =spans[i].getAttribute("id");
      if (id != null) refs["#"+id] = n;
    }
  }
  if (n == 0)
    noteholder.parentNode.removeChild(noteholder);
  else {
    // Process footnoterefs.
    for (i=0; i<spans.length; i++) {
      if (spans[i].className == "footnoteref") {
        var href = spans[i].getElementsByTagName("a")[0].getAttribute("href");
        href = href.match(/#.*/)[0];  // Because IE return full URL.
        n = refs[href];
        spans[i].innerHTML =
          "[<a href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
      }
    }
  }
}

}
/*]]>*/
</script>
</head>
<body>
<div id="header">
<h1>Just Annotate My Genome (JAMg) info</h1>
<span id="author">Alexie Papanicolaou</span><br />
<span id="email"><tt>&lt;<a href="mailto:alexie@butterflybase.org">alexie@butterflybase.org</a>&gt;</tt></span><br />
<span id="revnumber">version RC1,</span>
<span id="revdate">December 2013</span>
</div>
<div id="content">
<h2 id="_introduction">Introduction</h2>
<div class="sectionbody">
<div class="paragraph"><p>An increased number of genomes are being made public but few individual research are willing to take ownership of their own data. Indeed, the current model is for genome sequences to be handled by sequencing centers or large bioinformatic repositories (RefSeq or Ensembl). Even though using these widely used and standardized repositories and center is an excellent model to decrease the cost of completing a genome project, this comes at a cost. First, these groups have in-house pipelines built and customized for the projects that financially support them (i.e. small genomes of microbial human pathogens and largely fully complete genomes such as Drosophila or human) rather than say a highly polymorphic species from a natural ecosystem. Second, the lack of a robust funding model means that these repositories are do not have the resources to offer community-wide support and customization of a pipeline. Third, and perhaps most importantly, these centers and repositories usually lack the domain expertise associated with the biology of the species.</p></div>
<div class="paragraph"><p>For these (and perhaps others?) reasons, genome consortia that have access to genomicists (or PhD students and post-docs willing to learn) are either collaborating with bioinformatic laboratories or investing in their own annotation capability. This endeveavour has been greatly helped by the public availability of the tools used by the repositories and sequencing centers (e.g. GMOD, Ensembl and sequencing-center specific platforms such as those from the Broad Institute). The GMOD project specifically specializes in compact, user friendly solutions that <em>just_work</em>. For example, MAKER requires a few minutes of configuration to deliver a standardized annotation for gene models. At the other side of the spectrum, Ensembl delivers a comprehensive solution, database and informatic pipelines that - in the hands of a highly-trained bioinformatician - can deliver the same depth and level of annotation as that used by the EBI. There really is no a solution that fits in-between. There is also almost no software that also wishes to educate the user rather than offering a black box. Finally, there is no solution that we know of that can also functionally annotate the genome (a la BLAST2GO but free) and then link the concept of gene model (feature) annotation with functional annotation.</p></div>
<div class="paragraph"><p>The JAMg software was created to address the issue of creating gene models (feature annotation) and was built by <a href="http://tiny.cc/alexie_pap_csiro/">Alexie Papanicolaou</a> at the <a href="http://www.csiro.au/">Commonwealth Scientific and Industrial Research Organisation</a> (CSIRO) with some brilliant support from <a href="http://www.broadinstitute.org/~bhaas/">Brian Haas</a> at the <a href="http://broadinstitute.org">Broad Institute</a>. The software and manual are written so that to guide the annotation process so that users can follow the process closely. Even though with JAMg you will not need another genome annotation pipeline, JAMg does not aim to replace other genome annotation pipelines (e.g. each sequencing center has its own): it does aim to support nascent genome annotators and (ultimately) educate its users about genome annotation in general. As part of our Just_Annotate series, JAMg links with <a href="http://jamps.sourceforge.net/">JAMp</a> and WebApollo to provide a first solution for users wishing to go from genome assembly to deriving biological hypotheses.</p></div>
<h3 id="_genome_annotation">Genome annotation</h3><div style="clear:left"></div>
<div class="paragraph"><p>When genome biologists talk about annotation they can mean one of two things: <strong>feature annotation</strong> is identifying which parts of the genome are coding, the structure of the gene (where does each exon and ORF start &amp; end) and identifying any alternative splicing events. This allows us to build a database of transcripts and proteins for the entire species and the field of de-novo gene prediction (e.g. SNAP, Augustus, GeneMark etc) is a good example of the output. The second definition of annotation is <strong>functional annotation</strong>: what does a particular gene actually do. The Gene Ontology consortium is a good example of a group of scientists that deliver functional annotation knowledge: they read the literature and integrate it with any <em>in-silico</em> information that may exist. Alternatively, pure <em>in-silico</em> information (e.g. BLAST to a known protein) can deliver some information via inference.</p></div>
<div class="paragraph"><p><strong>Feature annotation</strong> is a mature field and an excellent example of the interdisciplinary use of maths, computer science and biology. There are a number of de-novo gene predictors (each with its strengths and weaknesses) and all rely on building a predictive model that is applied genome-wide (see <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3207268/">Haas et al 2011</a>). The use of models (such as Hidden Markov Models) trained by a handful of known sequences has performed exceptionally well as it seems that - as far as it is known - that species (or taxa) have a remarkable specific way of constructing their coding seqeunce (e.g. taxon-specific codon usage). However, it is naturally prone to a high rate of false positives (type II errors). The problem is a direct result of a large search space (millions of letters) that contain a dispersed and relatively small amount of signal (e.g. can be as low as 10 % of the total). Type I errors are less likely to occur because of the characteristics of coding sequences (we have a well understood structure such as a start, stop, no stop codons in the middle, exon/intron splice sites etc). Type II errors, however, occur when a stretch of stop-codon-free sequence looks coding but in fact it isn&#8217;t (either because it never was or it has become a pseudogene - worse still with modern genome sequencing: it could be a misassembly). Further, even if a gene model is predicted, it is not possible to identify the actual transcripts generated by this gene unless we have some external information: known transcripts from other species or transcriptomes. Finally, the ability to annotate UTR features is limited: the paucity of conserved signals and variability in the sequence complexity makes it difficult even though correct prediction of a long UTR (even if approximatelly correct) prevents from false positives being predicted in those regions.</p></div>
<div class="paragraph"><p>These two lines of evidence can help improve both the accuracy of genome annotation and construct the exact transcripts produced. The known transcripts can be from the same (e.g. <a href="http://trinityrnaseq.sourceforge.net/">de-novo transcript assembly</a>) or a related species (that has been annotated - but see below for <em>'propagation of annotation error</em>'). There are disadvantages to both approaches: the former relies on the transcript being actually expressed and correctly reconstructed, the latter requires the gene to be sufficiently conserved - a situation that will improve as we sequence and annotate more species. Most de-novo gene predictors do not or only mildly support such external evidence (they use it to add some weight on some predictions). A bright exception is AUGUSTUS which - if used correctly - has the in-built feature of using such lines of evidence to weight predictions, join exons into a single gene and identify alternative transcription events (via the use of <em>junction reads</em>).</p></div>
<div class="paragraph"><p>Further, even before the availability of such external evidence, an approach - used widely - is the use multiple de-novo gene predictors and use a voting mechanism (e.g. GLEAN or EvidenceModeller). These tools have the ability to bring together not only gene predictions but other lines of evidence (such as manually curated genes) and build a <em>consensus gene set</em>. Finally, even though we can use transcript evidence to guide gene prediction (and with AUGUSTUS also transcript prediction), once gene prediction is complete that transcript evidence must then be incorporated back into the consensus gene set, enabling us to identify more alternative transcripts and - now that we know what are the ORFs of the transcripts - accurately identify their UTR features. After manual inspection and correction of errors (i.e. curation), the biologist can subsequently repeat the entire process to deliver incrementally improved genome annotations. Alternatively, if multiple species are available, they can annotate an entire taxon and having purging obvious errors and low quality data, run a second annotation round informed with high quality models.</p></div>
<div class="literalblock">
<div class="content">
<pre><tt>_TODO_ improve clarity of some sentences</tt></pre>
</div></div>
<h3 id="_propagation_of_annotation_error_and_curation">Propagation of annotation error and curation</h3><div style="clear:left"></div>
<div class="literalblock">
<div class="content">
<pre><tt>(which happen often due to Type II error)... _TODO_</tt></pre>
</div></div>
<h3 id="_do_we_need_another_approach">Do we need another approach?</h3><div style="clear:left"></div>
<div class="paragraph"><p>Because of some characteristics, it can be easier to identify genes in compact genomes, those that lack introns (e.g. bacteria), lack duplicated in-tandem paralogues, exhibit little alternative splicing and have limited amount of repetitive sequence and within-species polymorphism (e.g. <em>Drosophila</em>, <em>C. elegans</em> and <em>Arabidobsis</em>) in relation to the average. Further, a very high quality genome assembly (ie. <em>finished</em>) will increase the chances of correct modelling and full ascertainment of a gene family. Finally, the sequencing of a diversity of relatively close related species (e.g. mammals) can vastly improve the accuracy since a conserved signal is more likely to be functional (such as coding) than not. For that reason, model species tend to be well annotated: they were picked to be models of the genome era because of some of the mentioned characteristics.</p></div>
<div class="paragraph"><p>However, researchers use myriads of species as models for specific questions in medicine &amp; physiology, ecology &amp; agriculture, evolution and other basic sciences. I call the former models <em>resource-rich models</em> (because they are good resources) and the latter <em>question-rich models</em> (because they can be used to answer particular questions). We have now reached the most exciting times were any question-rich model can be elevated to a resource-rich model using the vastly decreased costs of sequencing (one of the major secondary benefits of the Human Genome Project activities). Unfortunately, removing one bottleneck always exposes another. In this case the bottleneck is the bioinformatic analysis of these genomes, whether it is their assembly, annotation (both types) or other downstream experimentation.</p></div>
<div class="paragraph"><p>To respond to this bottleneck, sequencing centers have developed their own in-house pipelines. The <a href="http://gmod.org">GMOD</a> consortium has also develop a pipeline (MAKER) that standardizes and simplifies de-novo gene prediction. MAKER is exceptional in that it requires only a few minutes of a bioinformatician&#8217;s time (but weeks of computational time) and is the first software that brought gene prediction beyond the sequencing centers, standardized its use and linked with curation software such as Apollo. We found, however, that MAKER had some issues that did not suit our purpose: first because it didn&#8217;t offer much control on how the pipeline was run and it was too much of a black box (see commentary in Science <a href="http://www.sciencemag.org/content/340/6134/814">Joppa et al 2013</a>) so that without any incentitive to undestand gene prediction users rarely understood why certain tools were run (and therefore they couldn&#8217;t make a decision later on). More importantly, however, we found that MAKER doesn&#8217;t have an approach to train the de-novo gene predictors (the users has to do that on their own but that is not part of the pipeline and in practice MAKER users rarely do) and it made poor use of transcript evidence. Also when we first started working on this MAKER - because it was such a black box pipeline - was too slow and didn&#8217;t perform well on a HPC environment even though improvements have been made. In any case, basically we started with MAKER but I was dissatisfied with its performance for purely selfish reasons (basically it couldn&#8217;t annotate our gene family of interest&#8230;!). The other problem we came across was that most de-novo predictors (except Augustus) were poorly documented or, worse, unsupported. With poor documentation, it was excruciating trying to find out how to use them (every single software had its own file format!) and I don&#8217;t think anyone ought to have to go through that again.</p></div>
<div class="paragraph"><p>Having said that, there are some sequencing centers (such as the Broad) that have established complicated platforms, with extensive Quality control (QC). These however are really intended for internal use and we could find a single package that could deliver the best of both worlds without freaking out already over-worked bioinformaticians. So we set out to create a new approach to genome annotation that was educational, high quality, and contained as much capability was needed without being too much of a strain to bioinformaticians. We invested considerable time picking different software, but the modularity of our approach does not exclude incorporating any other approaches and software (including MAKER&#8217;s output). Our aim was <em>not</em> to create a point and click pipeline but a platform that even a bioinformatician - who is a capable user of bioinformatics but has never annotated - can pick up relatively quickly and deliver the highest possible quality of annotation.</p></div>
<div class="paragraph"><p>We call our platform Just_Annotate_My_genome (JAMg) and we hope you find it useful. In case you&#8217;re wondering, we considered following the model of ARGO but JAMg sounds better than JAMFg (and also I&#8217;ve run out of 50 cent coins).</p></div>
<h3 id="_intended_audience">Intended audience</h3><div style="clear:left"></div>
<div class="paragraph"><p>In particular, our audience are early career scientists who have <em>ownership</em> of their species and understand that annotation errors will have detrimental effects on their downstream experiments. Naturally, the output of any platform will depend on the data and effort invested but we feel that our approach will deliver the better possible results given the data.</p></div>
<div class="paragraph"><p>The software is not particularly useful to career bioinformaticians who have to churn dozens of genomes for a living but have no (scientific) interest in the underlying content. Other tools, such as MAKER or some sequencing center pipelines may be more appropriate high-throughput, one-size-fits-all methods.</p></div>
</div>
<h2 id="_just_annotate_my_genome">Just_Annotate_My_genome</h2>
<div class="sectionbody">
<h3 id="_what_is_jamg_for">What is JAMg for?</h3><div style="clear:left"></div>
<div class="paragraph"><p>JAMg is focused on delivering a process for feature annotation for all the species that are not <em>model species</em>. It links to another bit of software, <a href="http://jamps.sourceforge.net/">Just_Annotate_My_proteins (JAMp)</a> that functionally annotates proteins. JAMg is far from being a comprehensive solution (see Plugins needed) but it is a modular and an open-source software and like our other software we invite the community to provide not only feedback but also additional functionality and suggestions on how to improve it.</p></div>
<h3 id="_obtaining_jamg">Obtaining JAMg</h3><div style="clear:left"></div>
<div class="paragraph"><p>Currently only distributed through the author. It not published. We really appreciate any comments and feedback you can provide, so please do no matter how minor you think they are (e.g. typos in this document).</p></div>
<div class="paragraph"><p>We support only 64bit Linux (Ubuntu preferred) but users should feel free try other platforms. Repositories such as that from Ubuntu allow users to quickly install secondary software such as samtools.</p></div>
<div class="paragraph"><p>We found it painful to have to find and install all the various bit of software that are needed. For that reason, we distribute a number of 3rd party software (see 3rd_party) within the distribute to facilitate installation (often under the Agglomeration terms of GPL) but we assume that users have a 64bit Linux Operating System.</p></div>
<h3 id="_running_jamg">Running JAMg</h3><div style="clear:left"></div>
<div class="paragraph"><p>First, read the description of the JAMg <a href="procedure.html">procedure</a>.
Then, read the <a href="tutorial.html">tutorial</a>.</p></div>
<h3 id="_components_and_dependencies">Components and dependencies</h3><div style="clear:left"></div>
<div class="paragraph" id="software"><p>JAMg (and any gene annotation) really shines for species for which you have raw RNA-Seq (or other transcriptome) data but it can work without it too.</p></div>
<div class="paragraph"><p>These software are core to the way JAMg works:</p></div>
<div class="ulist"><ul>
<li>
<p>
<a href="ftp://toolkit.genzentrum.lmu.de/pub/HHblits">HHblits</a> used to identify transposable elements and domains not captured by transcriptomes (i.e. not expressed)
</p>
</li>
<li>
<p>
<a href="http://bioinf.uni-greifswald.de/augustus/">Augustus</a> our de-novo gene predictor of choice
</p>
</li>
<li>
<p>
<a href="http://www.repeatmasker.org‎">RepeatMasker</a>
</p>
</li>
<li>
<p>
<a href="http://apollo.berkeleybop.org/">WebApollo</a>
</p>
</li>
</ul></div>
<div class="paragraph"><p>And our other software:</p></div>
<div class="ulist"><ul>
<li>
<p>
<a href="http://pasa.sourceforge.net/">PASA2</a> (unpublished version 2): incorporate transcriptome information
</p>
</li>
<li>
<p>
<a href="http://trinityrnaseq.sourceforge.net/">Trinity RNA-Seq</a>: Arguably, the best de-novo RNA-seq assembler out there.
</p>
</li>
<li>
<p>
<a href="http://transdecoder.sourceforge.net">TransDecoder</a>: protein prediction from RNA-Seq assemblies, bundled in PASA2 so no need to install it separately.
</p>
</li>
<li>
<p>
<a href="http://aatpackage.sourceforge.net">aatpackage</a>: a simple and fast way to align proteins to a genome
</p>
</li>
<li>
<p>
<a href="http://evidencemodeler.sourceforge.net">EvidenceModeller</a>: constructs consensus gene sets
</p>
</li>
<li>
<p>
<a href="https://sourceforge.net/projects/justpreprocessmyreads">Just_Preprocess_My_reads</a>: a simple way of pre-processing and QCing RNA-Seq
</p>
</li>
<li>
<p>
<a href="http://jamps.sourceforge.net">JAMp</a>: functional annotation of your predicted proteins
</p>
</li>
</ul></div>
<div class="paragraph"><p>These dependencies exist due to the above software (on Ubuntu you can install them with apt-get):</p></div>
<div class="ulist"><ul>
<li>
<p>
<a href="http://www.molecularevolution.org/software/genomics/gmap">GMAP and GSNAP</a>
</p>
</li>
<li>
<p>
<a href="https://code.google.com/p/bedtools/‎">bedtools</a>
</p>
</li>
<li>
<p>
<a href="http://samtools.sourceforge.net/‎">samtools</a>
</p>
</li>
<li>
<p>
<a href="http://users.soe.ucsc.edu/~kent/src/blatSrc.zip">blat</a>
</p>
</li>
<li>
<p>
Tomcat and Apache webserver
</p>
</li>
</ul></div>
<div class="paragraph"><p>Further, we currently support these optional components:</p></div>
<div class="ulist"><ul>
<li>
<p>
<a href="http://exon.biology.gatech.edu/">GeneMark</a>: de-novo predictor without any training
</p>
</li>
<li>
<p>
<a href="http://genome.crg.es/software/geneid/">geneid</a>: no longer supported by authors
</p>
</li>
<li>
<p>
<a href="http://korflab.ucdavis.edu/software.html‎">snap</a>: SNAP from Ian Korf
</p>
</li>
<li>
<p>
<a href="http://ccb.jhu.edu/software/glimmerhmm/‎">Glimmer</a>: no longer supported by authors
</p>
</li>
</ul></div>
<div class="paragraph"><p>All of the above software are installed with</p></div>
<div class="listingblock">
<div class="content"><!-- Generator: GNU source-highlight 3.1.3
by Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite -->
<pre><tt>make all</tt></pre></div></div>
<div class="paragraph"><p>except for blat, WebApollo, Tomcat, Apache and the optional <a href="http://topaz.gatech.edu/license_download.cgi">GeneMarkES</a>. You will need to manually install those. You also agree that it is your responsibility to abide by the various license agreements.</p></div>
<div class="paragraph"><p>See 3rd_party/WebApollo for our webservices, you will need a specific version of WebApollo.</p></div>
<h3 id="_wishlist">Wishlist</h3><div style="clear:left"></div>
<div class="ulist"><ul>
<li>
<p>
Implementing GMAP on top of aatpackage for identifying high quality subset.
</p>
</li>
<li>
<p>
Tutorials and videos
</p>
</li>
</ul></div>
<h3 id="_the_team">The Team</h3><div style="clear:left"></div>
<div class="paragraph"><p>Alexie Papanicolaou<sup>1</sup> and Brian Haas<sup>2</sup><br />
   <sup>1</sup> CSIRO Ecosystem Sciences, GPO 1700, Canberra 2601, Australia<br />
   <sup>2</sup> The Broad Institute, Cambridge, MA, USA<br />
   <a href="mailto:alexie@butterflybase.org">alexie@butterflybase.org</a></p></div>
<div class="literalblock">
<div class="content">
<pre><tt>Copyright CSIRO 2013-2014.
This software is released under the Mozilla Public License v.2. You can find the terms and conditions at http://www.mozilla.org/MPL/2.0.
It is provided "as is" without warranty of any kind.</tt></pre>
</div></div>
</div>
</div>
<div id="footnotes"><hr /></div>
<div id="footer">
<div id="footer-text">
Version RC1<br />
Last updated 2014-01-03 23:29:03 EST
</div>
</div>
</body>
</html>
